pkg/controllers/upgrade_reconcile.go

```diff
@@ -0,0 +1,271 @@
+/*
+ (c) Copyright [2021] Micro Focus or one of its affiliates.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ You may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	vapi "github.com/vertica/vertica-kubernetes/api/v1beta1"
+	"github.com/vertica/vertica-kubernetes/pkg/cmds"
+	"github.com/vertica/vertica-kubernetes/pkg/events"
+	"github.com/vertica/vertica-kubernetes/pkg/names"
+	"github.com/vertica/vertica-kubernetes/pkg/status"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+// UpgradeReconciler will update the status field of the vdb.
+type UpgradeReconciler struct {
+	VRec              *VerticaDBReconciler
+	Log               logr.Logger
+	Vdb               *vapi.VerticaDB // Vdb is the CRD we are acting on.
+	PRunner           cmds.PodRunner
+	PFacts            *PodFacts
+	ContinuingUpgrade bool // true if UpdateInProgress was already set upon entry
+}
+
+// MakeUpgradeReconciler will build an UpgradeReconciler object
+func MakeUpgradeReconciler(vdbrecon *VerticaDBReconciler, log logr.Logger,
+	vdb *vapi.VerticaDB, prunner cmds.PodRunner, pfacts *PodFacts) ReconcileActor {
+	return &UpgradeReconciler{VRec: vdbrecon, Log: log, Vdb: vdb, PRunner: prunner, PFacts: pfacts}
+}
+
+// Reconcile will update the status of the Vdb based on the pod facts
+func (u *UpgradeReconciler) Reconcile(ctx context.Context, req *ctrl.Request) (ctrl.Result, error) {
+	if err := u.PFacts.Collect(ctx, u.Vdb); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if ok, err := u.isUpgradeNeeded(ctx); !ok || err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Functions to perform upgrade processing.  Order matters.
+	funcs := []func(context.Context) (ctrl.Result, error){
+		// Initiate an upgrade by setting condition and event recording
+		u.startUpgrade,
+		// Do a clean shutdown of the cluster
+		u.stopCluster,
+		// Delete the sts objects.  We don't rely on k8s rolling upgrade.
+		// Everything must be destroyed then regenerated.
+		u.deleteStatefulSets,
+		// Create the sts object with the new image name.
+		u.recreateStatefulSets,
+		// Start up vertica in each pod.
+		u.restartCluster,
```

Comment:
Do we check if all the images are of the same version in/before this step?

Reply:
No, we don't.  But we are using the same image for each statefulset, so it will be the same version.